### ISSUE TITLE
Add Analyzer warning when an identifier in the local scope is overriding a type from the outer scope

### ISSUE THREAD

**Issue by michael-hawker** [dotnet/roslyn#68764]
**Brief description:**

This analyzer would help developers understand when they're trying to use a `type` within a scope that has been overridden by some other identifier (most likely a member of the current class or local scope). It should most likely be flagged on the definition of the identifier locally that has been given the same name the causes the resulting conflict later on in the program that is not the source of the underlying issue.

> CS###: 'identifier' exists in both 'parent type/namespace' and as 'type' from 'namespace'.
>
> This warning occurs in a situation when an identifier has been defined, usually a member within the current scope, that is given the same name as an available type in the outer scope, which produces ambiguity.

These types of issues can be hard and time consuming to debug as other analyzers generally flag problems with the code later where it's trying to be used, like [CS0117](https://learn.microsoft.com/dotnet/csharp/misc/cs0117), which points developers in the wrong direction where they focus on where they're trying to implement things and checking docs for types to ensure properties exist vs. that there's actually something else they've defined in their program which is disrupting their ability to use a type they're expecting.

It'd be much clearer to get a message above, akin to what happens with conflicting types in the same namespace ala [CS0433](https://learn.microsoft.com/dotnet/csharp/language-reference/compiler-messages/cs0433) to help the developer realize there's a conflict.

**Languages applicable:**

Probably C# only, but not familiar enough with VB anymore to know if the same would apply.

**Code example that the analyzer should report:**

A small code snippet that describes a case that the analyzer should report.

```cs
// Suppose you have an enum/type/class or something defined somewhere (or are using from a platform API or library)
public enum Orientation {
    Horiztonal,
    Vertical,
}

// Elsewhere you define a class that has a property with the same name as the type, in our case this was generated from a source generator so this issue was super not obvious looking at our code...
public partial class MyClass
{
    public string Orientation { get; set; } // New Analyzer would flag this line with the warning when used in a way that would be conflated with the type of the same name (see below)
}

// In that same class with the property defined, you try and use your type, as one would do... however, this doesn't work as the compiler is now confused by the local property defined as `Orientation` with a different type (string)... but there's no warning or analyzer to tell you that. Instead, you get things about static methods CS0120 or if you remove the 'static' from the method you get CS0117 instead...
public partial class MyClass
{
    public static Orientation ConvertStringToOrientation(string orientation) => orientation switch
    {
        "Vertical" => Orientation.Vertical,
        "Horizontal" => Orientation.Horizontal,
        _ => throw new System.NotImplementedException(),
    };
}
```

**Additional information:**

- This came up because we had a source generator generating are property with the same name as an enum type we were trying to use, but with a different type, `string`. It was not obvious that this was the issue at all with a naming conflict between the local identifier and our enum type in the parent scope we wanted to use. We were getting things like [CS0120](https://learn.microsoft.com/dotnet/csharp/language-reference/compiler-messages/cs0120) related to the method and then if we made it not `static` a [CS0117](https://learn.microsoft.com/dotnet/csharp/misc/cs0117) about `Horizontal` not being a property on `string`... which finally at least clued us into a bit more about what had happened.
  
  If the compiler just told us this though or warned us that something in the local scope was overriding the type definition, this would have saved us a good 15 minutes on a group call, let alone how much time the developer was confused with this beforehand as this was based on an existing example code sample that worked fine (just had a different generated property name).

- These types of scenarios seem common and something that the compiler could help developers with to better understand when they have conflicting definitions under the same scope. See this decade old related [stack overflow post as well](https://stackoverflow.com/questions/11379166/string-does-not-contain-a-definition-for).

- This is related to other conflict messages like [CS0433](https://learn.microsoft.com/dotnet/csharp/language-reference/compiler-messages/cs0433), [CS0434](https://learn.microsoft.com/dotnet/csharp/misc/cs0434), [CS0438](https://learn.microsoft.com/dotnet/csharp/misc/cs0438) which the compiler identifiers and helps easily inform the developer that there is an overloaded identifier in the current scope.

**Open questions**

- [ ] The [Framework Design Guidelines - Names of Type Members](https://learn.microsoft.com/dotnet/standard/design-guidelines/names-of-type-members#names-of-properties) explicitly suggests that giving the property the same name as the type is OK. This is common in UI orientated classes. However, in this case the property name and type usually have the same name, generally these are enums and the compiler/VS seems better about suggesting and not conflicting the proper usage there. Therefore, this seems to be more of an issue when there are mismatching types between the property name and the type name? But not sure if that's a stead-fast distinction in all cases. 

  So, what is the appropriate frequency/test needed to determine if a case should be flagged as suppressing all properties in UI controls would be excessive... Can it be indicated if the compiler gets confused later when by usage, as in a UI class there's usually distinctions and this scenario developers face generally isn't an issue to the same degree...

- [ ] In our scenario, the issue came from a generated property. Normally some of those analyzers get suppressed, right? Though in this case it'd be more important to notify the developer about it? So not sure how that would get resolved... As it's not a figment of how the code was generated but instead the impact it has on your other code the developer is writing...

**Documentation requirements:**

When this analyzer is implemented, it must be documented by following the steps at [Documentation for IDE CodeStyle analyzers](https://github.com/dotnet/roslyn/blob/main/docs/contributing/Documentation%20for%20IDE%20CodeStyle%20analyzers.md).

**Comment by CyrusNajmabadi** [comment]
MOving back to roslyn and giving to compiler to see about improvements in diagnostic clarity here :)
